File: UrbanPy/Backend_fastAPI/src/core/engine/contract.py
"""
Contrat du moteur pur : la représentation compacte (entiers, booléens, aucun texte) d'une partie, que tout moteur —
le moteur Python de référence comme un moteur compilé — doit savoir jouer à l'identique. Deux moitiés :
  - le deck, immuable pendant la partie : 8 cartes compilées, dont les capacités sont des indices dans un vocabulaire
    figé et des masques de bits ;
  - l'état, qui change à chaque round : round, premier joueur, vies, pillz, cartes jouées, effets persistants (dans
    l'ordre d'activation : le niveau 4 les applique dans cet ordre), dernier round (lu par Revenge / Confidence / After).
Un état contient exactement ce que process_round lit, rien de plus : test_engine_contract rejoue des parties depuis la
forme compacte et exige le même résultat qu'avec les objets d'origine.
Le vocabulaire est figé ici (et non dérivé du parseur) pour que ses indices soient stables : un moteur compilé les
reprend tels quels. Ajouter une valeur = l'ajouter en fin de tuple.
"""
from dataclasses import dataclass
from typing import NamedTuple, Optional, Tuple

from src.core.domain.capacity import Capacity
from src.core.domain.card import Card
from src.core.domain.effect import PersistentEffect
from src.core.domain.game import Game
from src.core.domain.player import Player
from src.core.domain.round import Round

HOWS = ("", "Protection", "brawl", "cancel", "copy", "counter_attack", "degrowth", "equalizer", "exchange", "growth",
        "impose", "limitless", "nb_dam_opp", "nb_damage", "nb_life_left", "nb_life_lost", "nb_pillz_left",
        "nb_pillz_lost", "nb_pow_opp", "stop", "support", "tie_break", "tune_out")
TYPES = ("ability", "attack", "bonus", "combust", "consume", "counter_attack", "damage", "dope", "heal", "infiltrated",
         "ko", "life", "limitless", "pillz", "poison", "power", "reanimate", "recover", "regen", "repair", "tie_break",
         "toxine", "tune_out")
CONDITIONS = ("after", "asymmetry", "backlash", "bet", "confidence", "courage", "defeat", "disunion", "infiltrated",
              "killshot", "perfect", "reprisal", "revenge", "stop", "symmetry", "team", "unison", "versus",
              "victory_defeat")
CLAN_CONDITIONS = ("versus", "after", "infiltrated")   # conditions paramétrées par une liste de clans
TARGETS = ("ally", "enemy", "both")
CLANS = ("All Stars", "Bangers", "Berzerk", "Cosmohnuts", "Dominion", "Fang Pi Clang", "Freaks", "Frozn", "GHEIST",
         "GhosTown", "Hive", "Huracan", "Jungo", "Junkz", "Komboka", "La Junta", "Leader", "Montana", "Nightmare",
         "Oblivion", "Oculus", "Paradox", "Piranas", "Pussycats", "Raptors", "Rescue", "Riots", "Roots", "Sakrohm",
         "Sentinel", "Skeelz", "Tolvack", "Ulu Watu", "Uppers", "Vortex", "Zenith")
EFFECT_KINDS = PersistentEffect.KINDS


@dataclass(frozen=True)
class CompiledCapacity:
    how: int              # indice dans HOWS
    target: int           # indice dans TARGETS
    types: int            # masque de bits sur TYPES
    value: int
    borne: int            # -1 = sans borne
    conditions: int       # masque de bits sur CONDITIONS
    bet_over: int = 0     # « Bet > N » : N, si le bit « bet » est levé et bet_under vaut 0
    bet_under: int = 0    # « Bet < N » : N (jamais 0 : une mise n'est pas négative)
    clans: int = 0        # masque de bits sur CLANS de la condition versus / after / infiltrated (une seule par capacité)


@dataclass(frozen=True)
class CompiledCard:
    name: str             # identité, pour lire un corpus ; le moteur ne la lit pas
    stars: int
    clan: int             # indice dans CLANS
    power: int
    damage: int
    ability: Optional[CompiledCapacity]
    bonus: Optional[CompiledCapacity]


@dataclass(frozen=True)
class Deck:
    ally: Tuple[CompiledCard, ...]
    enemy: Tuple[CompiledCard, ...]


@dataclass(frozen=True)
class PlayerState:
    life: int
    pillz: int
    played: Tuple[bool, ...]                    # une entrée par carte de la main
    effects: Tuple[Tuple[int, int, int], ...]   # (indice dans EFFECT_KINDS, value, borne), dans l'ordre d'activation


@dataclass(frozen=True)
class State:
    nb_turn: int
    ally_first: bool                            # Game.turn : l'allié joue en premier ce round
    ally: PlayerState
    enemy: PlayerState
    last_round: Optional[Tuple[int, int, bool]] # (carte alliée, carte ennemie, l'allié a gagné) ; None au round 1


class Action(NamedTuple):
    card: int         # indice dans la main
    pillz: int        # pillz_fight : 1 + mise (la pillz gratuite comprise)
    fury: bool


# --- Vocabulaire <-> masques -------------------------------------------------------------------

def _mask(vocabulary: tuple, names) -> int:
    mask = 0
    for name in names:
        mask |= 1 << vocabulary.index(name)
    return mask


def _names(vocabulary: tuple, mask: int) -> list:
    return [name for index, name in enumerate(vocabulary) if mask >> index & 1]


# --- Capacité ---------------------------------------------------------------------------------

def compile_capacity(capacity: Optional[Capacity]) -> Optional[CompiledCapacity]:
    if capacity is None:
        return None
    conditions, bet_over, bet_under, clans = 0, 0, 0, 0
    for condition in capacity.effect_conditions:
        name, _, parameter = condition.partition(":")
        if name.startswith("bet"):                       # « bet>N », « bet<N » (forme historique « bet N » = « bet>N »)
            name, parameter = "bet", condition[3:].strip()
            if parameter.startswith("<"):
                bet_under = int(parameter[1:])
            else:
                bet_over = int(parameter.lstrip(">").strip())
        elif parameter:
            clans = _mask(CLANS, parameter.split("|"))
        conditions |= 1 << CONDITIONS.index(name)
    return CompiledCapacity(how=HOWS.index(capacity.how), target=TARGETS.index(capacity.target),
                            types=_mask(TYPES, capacity.types), value=capacity.value, borne=capacity.borne,
                            conditions=conditions, bet_over=bet_over, bet_under=bet_under, clans=clans)


def capacity_from_compiled(compiled: Optional[CompiledCapacity]) -> Optional[Capacity]:
    if compiled is None:
        return None
    conditions = []
    for name in _names(CONDITIONS, compiled.conditions):
        if name == "bet":
            conditions.append(f"bet<{compiled.bet_under}" if compiled.bet_under else f"bet>{compiled.bet_over}")
        elif name in CLAN_CONDITIONS:
            conditions.append(name + ":" + "|".join(_names(CLANS, compiled.clans)))
        else:
            conditions.append(name)
    return Capacity(target=TARGETS[compiled.target], types=_names(TYPES, compiled.types), value=compiled.value,
                    borne=compiled.borne, how=HOWS[compiled.how], effect_conditions=conditions)


# --- Carte ------------------------------------------------------------------------------------

def compile_card(card: Card) -> CompiledCard:
    return CompiledCard(name=card.name, stars=card.stars, clan=CLANS.index(card.faction), power=card.power,
                        damage=card.damage, ability=compile_capacity(card.ability), bonus=compile_capacity(card.bonus))


def card_from_compiled(compiled: CompiledCard, played: bool) -> Card:
    """Carte du moteur Python bâtie sur la forme compilée seule (sans relire les données officielles)."""
    card = Card.__new__(Card)
    card.name = compiled.name
    card.faction = CLANS[compiled.clan]
    card.starOff = 0
    card.stars = compiled.stars
    card.power = compiled.power
    card.damage = compiled.damage
    card.ability = capacity_from_compiled(compiled.ability)
    card.bonus = capacity_from_compiled(compiled.bonus)
    card.image = card.clan_image = ""
    card.ability_description = card.bonus_description = ""
    card.power_fight = card.damage_fight = card.pillz_fight = card.attack = 0
    card.ability_fight = card.bonus_fight = card.leader_fight = None
    card.fury = card.win = False
    card.played = played
    card.cancelled_modifs = set()
    return card


# --- Partie <-> (deck, état) ------------------------------------------------------------------

def deck_from_game(game: Game) -> Deck:
    return Deck(ally=tuple(compile_card(card) for card in game.ally.cards),
                enemy=tuple(compile_card(card) for card in game.enemy.cards))


def _player_state(player: Player) -> PlayerState:
    return PlayerState(life=player.life, pillz=player.pillz, played=tuple(card.played for card in player.cards),
                       effects=tuple((EFFECT_KINDS.index(effect.kind), effect.value, effect.borne)
                                     for effect in player.effect_list))


def state_from_game(game: Game) -> State:
    last = game.history[-1] if game.history else None
    return State(nb_turn=game.nb_turn, ally_first=game.turn, ally=_player_state(game.ally),
                 enemy=_player_state(game.enemy),
                 last_round=None if last is None else (last.ally.card_index, last.enemy.card_index, last.ally.win))


def _player(name: str, cards: Tuple[CompiledCard, ...], state: PlayerState) -> Player:
    return Player(name=name, life=state.life, pillz=state.pillz,
                  cards=[card_from_compiled(card, played) for card, played in zip(cards, state.played)],
                  effect_list=[PersistentEffect(EFFECT_KINDS[kind], value, borne) for kind, value, borne in state.effects])


def game_from_state(deck: Deck, state: State) -> Game:
    """Partie jouable par process_round ; l'historique est réduit au dernier round, seul lu par le moteur."""
    history = []
    if state.last_round is not None:
        ally_index, enemy_index, ally_won = state.last_round
        last = Round()
        last.ally.card_index, last.ally.win = ally_index, ally_won
        last.enemy.card_index, last.enemy.win = enemy_index, not ally_won
        history.append(last)
    return Game(nb_turn=state.nb_turn, turn=state.ally_first, ally=_player("ally", deck.ally, state.ally),
                enemy=_player("enemy", deck.enemy, state.enemy), history=history)


# --- JSON (l'écriture est dataclasses.asdict ; les tuples y deviennent des listes) ------------

def _capacity_from_dict(fields: Optional[dict]) -> Optional[CompiledCapacity]:
    return None if fields is None else CompiledCapacity(**fields)


def _card_from_dict(fields: dict) -> CompiledCard:
    return CompiledCard(**{**fields, "ability": _capacity_from_dict(fields["ability"]),
                           "bonus": _capacity_from_dict(fields["bonus"])})


def deck_from_dict(fields: dict) -> Deck:
    return Deck(ally=tuple(map(_card_from_dict, fields["ally"])), enemy=tuple(map(_card_from_dict, fields["enemy"])))


def _player_state_from_dict(fields: dict) -> PlayerState:
    return PlayerState(life=fields["life"], pillz=fields["pillz"], played=tuple(fields["played"]),
                       effects=tuple(tuple(effect) for effect in fields["effects"]))


def state_from_dict(fields: dict) -> State:
    last_round = fields["last_round"]
    return State(nb_turn=fields["nb_turn"], ally_first=fields["ally_first"],
                 ally=_player_state_from_dict(fields["ally"]), enemy=_player_state_from_dict(fields["enemy"]),
                 last_round=None if last_round is None else tuple(last_round))

File: UrbanRust/src/contract.rs
//! L'état compact d'une partie, transcrit de `UrbanPy/Backend_fastAPI/src/core/engine/contract.py`.
//!
//! Deux moitiés, comme côté Python : le **deck**, immuable pendant la partie, et l'**état**, qui change à chaque
//! round. Tout est de taille fixe et `Copy` — aucun tas, aucune indirection : l'IA copie, hache et compare des états
//! par millions, c'est le seul endroit où la représentation décide de la vitesse.
//!
//! Deux bornes justifient les tableaux fixes :
//!   - une main fait exactement 4 cartes (validé par `game_schemas.py`) ;
//!   - un joueur porte au plus 8 effets persistants, un par sorte : `register_persistent_effect` remplace l'effet
//!     de même sorte au lieu de l'empiler.

use crate::vocabulary::EFFECT_KINDS;

pub const HAND_SIZE: usize = 4;
pub const MAX_EFFECTS: usize = EFFECT_KINDS.len();

/// Capacité compilée : plus un mot, rien que des indices dans le vocabulaire et des masques de bits.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct CompiledCapacity {
    pub how: u8,    // indice dans HOWS
    pub target: u8, // indice dans TARGETS
    pub types: u32, // masque de bits sur TYPES
    pub value: i16,
    pub borne: i16,      // -1 = sans borne
    pub conditions: u32, // masque de bits sur CONDITIONS
    pub bet_over: u8,    // « Bet > N » : N, si le bit « bet » est levé et bet_under vaut 0
    pub bet_under: u8,   // « Bet < N » : N (jamais 0 : une mise n'est pas négative)
    pub clans: u64,      // masque de bits sur CLANS de la condition versus / after / infiltrated
}

/// Carte compilée. Le nom n'y est pas : il sert à lire un corpus, jamais à jouer un round.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct CompiledCard {
    pub stars: u8,
    pub clan: u8, // indice dans CLANS
    pub power: i16,
    pub damage: i16,
    pub ability: Option<CompiledCapacity>,
    pub bonus: Option<CompiledCapacity>,
}

/// Les huit cartes de la partie et la situation de départ de chaque camp, lue par « Par Vie / Pillz perdue ».
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct Deck {
    pub ally: [CompiledCard; HAND_SIZE],
    pub enemy: [CompiledCard; HAND_SIZE],
    pub ally_start: (i16, i16), // vie, pillz
    pub enemy_start: (i16, i16),
}

#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct Effect {
    pub kind: u8, // indice dans EFFECT_KINDS
    pub value: i16,
    pub borne: i16, // -1 = sans borne
}

impl Effect {
    /// Case vide d'un tableau d'effets : `kind` hors vocabulaire, pour que deux états égaux se hachent pareil.
    const EMPTY: Effect = Effect { kind: u8::MAX, value: 0, borne: 0 };
}

#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct PlayerState {
    pub life: i16,
    pub pillz: i16,
    pub played: u8, // bit i levé = la carte i de la main a déjà été jouée
    effect_count: u8,
    effects: [Effect; MAX_EFFECTS], // dans l'ordre d'activation : le niveau 4 les applique dans cet ordre
}

impl PlayerState {
    pub fn new(life: i16, pillz: i16, played: u8) -> Self {
        PlayerState {
            life,
            pillz,
            played,
            effect_count: 0,
            effects: [Effect::EMPTY; MAX_EFFECTS],
        }
    }

    pub fn has_played(&self, card: usize) -> bool {
        self.played >> card & 1 == 1
    }

    pub fn set_played(&mut self, card: usize) {
        self.played |= 1 << card;
    }

    pub fn effects(&self) -> &[Effect] {
        &self.effects[..self.effect_count as usize]
    }

    /// Enregistre l'effet en remplaçant celui de même sorte s'il existe, qui perd sa place dans l'ordre
    /// d'activation (`register_persistent_effect`, côté Python).
    pub fn register(&mut self, effect: Effect) {
        let mut kept = 0;
        for index in 0..self.effect_count as usize {
            if self.effects[index].kind != effect.kind {
                self.effects[kept] = self.effects[index];
                kept += 1;
            }
        }
        self.effects[kept] = effect;
        self.effect_count = kept as u8 + 1;
    }
}

/// Ce que le moteur lit du round précédent : les deux cartes jouées et le vainqueur (Revenge / Confidence / After).
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct LastRound {
    pub ally_card: u8,
    pub enemy_card: u8,
    pub ally_won: bool,
}

#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct State {
    pub nb_turn: u8,
    pub ally_first: bool, // l'allié joue en premier ce round
    pub ally: PlayerState,
    pub enemy: PlayerState,
    pub last_round: Option<LastRound>, // None au round 1
}

/// Un coup : la carte de la main, `pillz_fight` (1 + la mise, la pillz gratuite comprise) et la fury.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct Action {
    pub card: u8,
    pub pillz: i16,
    pub fury: bool,
}

/// Valeurs de combat d'une carte à la fin du round : ce que le corpus exige de retrouver.
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct SideOutcome {
    pub power: i16,
    pub damage: i16,
    pub attack: i16,
    pub win: bool,
}

#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct Outcome {
    pub ally: SideOutcome,
    pub enemy: SideOutcome,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn effect(kind: u8, value: i16) -> Effect {
        Effect { kind, value, borne: -1 }
    }

    #[test]
    fn un_effet_remplace_celui_de_meme_sorte_et_passe_en_dernier() {
        let mut player = PlayerState::new(12, 12, 0);
        player.register(effect(0, 2));
        player.register(effect(1, 3));
        player.register(effect(0, 5));

        assert_eq!(player.effects(), &[effect(1, 3), effect(0, 5)]);
    }

    /// La mémoïsation de l'IA hache l'état entier, cases d'effets libres comprises : deux joueurs arrivés au même
    /// jeu d'effets par des chemins différents doivent se hacher pareil, sans quoi la table double les entrées.
    #[test]
    fn deux_chemins_vers_les_memes_effets_donnent_le_meme_etat() {
        use std::collections::hash_map::DefaultHasher;
        use std::hash::{Hash, Hasher};

        let mut direct = PlayerState::new(12, 12, 0);
        direct.register(effect(3, 1));
        direct.register(effect(7, 4));

        let mut apres_remplacement = PlayerState::new(12, 12, 0);
        apres_remplacement.register(effect(7, 9));
        apres_remplacement.register(effect(3, 1));
        apres_remplacement.register(effect(7, 4));

        assert_eq!(direct, apres_remplacement);

        let empreinte = |state: &PlayerState| {
            let mut hasher = DefaultHasher::new();
            state.hash(&mut hasher);
            hasher.finish()
        };
        assert_eq!(empreinte(&direct), empreinte(&apres_remplacement));
    }

    #[test]
    fn les_cartes_jouees_tiennent_dans_le_masque() {
        let mut player = PlayerState::new(12, 12, 0);
        player.set_played(2);
        assert!(player.has_played(2));
        assert!(!player.has_played(0));
    }
}
